tokio/src/io/uring/read.rs

```diff
@@ -0,0 +1,54 @@
+use crate::runtime::driver::op::{CancelData, Cancellable, Completable, CqeResult, Op};
+
+use io_uring::{opcode, types};
+use std::io::{self, Error};
+use std::os::fd::{AsRawFd, OwnedFd};
+
+#[derive(Debug)]
+pub(crate) struct Read {
+    fd: OwnedFd,
+    buf: Vec<u8>,
+}
+
+impl Completable for Read {
+    type Output = (io::Result<u32>, OwnedFd, Vec<u8>);
+
+    fn complete(self, cqe: CqeResult) -> Self::Output {
+        (cqe.result, self.fd, self.buf)
+    }
```

Comment:
Moving the `set_len` here seems to simplify the other file quite a bit.
```suggestion
    type Output = (io::Result<u32>, OwnedFd, Vec<u8>);

    fn complete(mut self, cqe: CqeResult) -> Self::Output {
        if let Ok(len) = cqe.result {
            let new_len = self.buf.len() + len as usize;
            unsafe { self.buf.set_len(new_len) };
        }

        (cqe.result, self.fd, self.buf)
    }
```